core/src/smartcontracts/wasm.rs

```diff
@@ -69,25 +69,80 @@ pub mod import {
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
     /// Engine or linker could not be created
-    #[error("Runtime initialization failure: {0}")]
+    #[error("Runtime initialization failure")]
     Initialization(#[source] eyre::Report),
     /// Module could not be compiled or instantiated
-    #[error("Module instantiation failure: {0}")]
+    #[error("Module instantiation failure")]
     Instantiation(#[source] eyre::Report),
     /// Expected named export not found in module
-    #[error("Named export not found: {0}")]
+    #[error("Named export not found")]
     ExportNotFound(#[source] eyre::Report),
+    /// Error exporting memory from wasm to iroha
+    #[error("Failed to export memory")]
+    MemoryExport(#[from] MemoryExportError),
     /// Call to the function exported from module failed
-    ///
-    /// In Wasmtime v0.33, can also mean that max linear memory was
-    /// consumed
-    #[error("Exported function call failed: {0}")]
-    ExportFnCall(#[from] Trap),
+    #[error("Exported function call failed")]
+    ExportFnCall(#[from] ExportFnCallError),
     /// Error during decoding object with length prefix
-    #[error("Failed to decode object from bytes with length prefix: {0}")]
+    #[error("Failed to decode object from bytes with length prefix")]
     Decode(#[source] eyre::Report),
 }
 
+/// Memory export error
+#[derive(Debug, Copy, Clone, thiserror::Error)]
+pub enum MemoryExportError {
+    /// Memory export not found
+    #[error("`{}` not found", export::WASM_MEMORY_NAME)]
+    NotFound,
+    /// Memory export is not a memory
+    #[error("`{}` is not a memory", export::WASM_MEMORY_NAME)]
+    NotAMemory,
+}
+
+/// Exported function call error
+#[derive(Debug, thiserror::Error)]
+pub enum ExportFnCallError {
+    /// Failed to execute something on the host side
+    #[error("Failed to execute operation")]
+    HostExecution(#[source] eyre::Report),
+    /// Stack overflow, heap overflow or other limits exceeded
+    #[error("Execution limits exceeded")]
+    ExecutionLimitsExceeded(#[source] eyre::Report),
+    /// Some integer error like division by zero or overflow
+    #[error("Numerical error")]
+    Integer(#[source] eyre::Report),
+    /// Other kind of trap
+    #[error("Other")]
+    Other(#[source] Trap),
```

Comment:
```suggestion
    Other(#[source] eyre::Report),
```

please don't expose `Trap`

Reply:
It was exposed before, but ok, I'll change it